File: src/lib/research/jurisdictionRegistry.ts
// California jurisdiction registry. Resolves WHICH local authority's rules apply
// to a facility by location. Air districts and regional water boards are FINITE
// and fully enumerated here (verified against arb.ca.gov / waterboards.ca.gov).
//
// Honest limits, by design:
//  - Five+ counties are split across air districts by sub-county geometry (LA,
//    San Bernardino, Riverside, Kern, Sonoma, Solano). resolveAirDistrict flags
//    these as needsGeometry: the caller must resolve by point-in-polygon, not by
//    county name alone.
//  - CUPAs (~81) and per-county/city fire+building codes are NOT enumerated here.
//    The fire/building layer is effectively unbounded (~480 cities x 3-year code
//    cycles) and must be resolved per-jurisdiction at runtime, not pre-tabled.

export type AirDistrict = { id: string; name: string; counties: string[]; website: string };
export type RegionalWaterBoard = { id: string; name: string; coverage: string };

export const AIR_DISTRICTS: AirDistrict[] = [
  { id: "amador-county-apcd", name: "Amador County APCD", counties: ["Amador"], website: "https://www.amadorgov.org" },
  { id: "antelope-valley-aqmd", name: "Antelope Valley AQMD", counties: ["Los Angeles"], website: "https://www.avaqmd.ca.gov" },
  { id: "bay-area-aqmd", name: "Bay Area AQMD", counties: ["Alameda", "Contra Costa", "Marin", "Napa", "San Francisco", "San Mateo", "Santa Clara", "Solano", "Sonoma"], website: "https://www.baaqmd.gov" },
  { id: "butte-county-aqmd", name: "Butte County AQMD", counties: ["Butte"], website: "https://www.bcaqmd.org" },
  { id: "calaveras-county-apcd", name: "Calaveras County APCD", counties: ["Calaveras"], website: "https://ema.calaverasgov.us" },
  { id: "colusa-county-apcd", name: "Colusa County APCD", counties: ["Colusa"], website: "https://www.countyofcolusaca.gov" },
  { id: "eastern-kern-apcd", name: "Eastern Kern APCD", counties: ["Kern"], website: "https://www.kernair.org" },
  { id: "el-dorado-county-aqmd", name: "El Dorado County AQMD", counties: ["El Dorado"], website: "https://www.eldoradocounty.ca.gov" },
  { id: "feather-river-aqmd", name: "Feather River AQMD", counties: ["Sutter", "Yuba"], website: "https://www.fraqmd.org" },
  { id: "glenn-county-apcd", name: "Glenn County APCD", counties: ["Glenn"], website: "https://www.countyofglenn.net" },
  { id: "great-basin-unified-apcd", name: "Great Basin Unified APCD", counties: ["Alpine", "Inyo", "Mono"], website: "https://www.gbuapcd.org" },
  { id: "imperial-county-apcd", name: "Imperial County APCD", counties: ["Imperial"], website: "https://www.imperialcounty.net" },
  { id: "lake-county-aqmd", name: "Lake County AQMD", counties: ["Lake"], website: "https://www.lcaqmd.net" },
  { id: "lassen-county-apcd", name: "Lassen County APCD", counties: ["Lassen"], website: "https://www.lassenair.org" },
  { id: "mariposa-county-apcd", name: "Mariposa County APCD", counties: ["Mariposa"], website: "https://www.mariposacounty.org" },
  { id: "mendocino-county-aqmd", name: "Mendocino County AQMD", counties: ["Mendocino"], website: "https://www.co.mendocino.ca.us" },
  { id: "modoc-county-apcd", name: "Modoc County APCD", counties: ["Modoc"], website: "https://www.co.modoc.ca.us" },
  { id: "mojave-desert-aqmd", name: "Mojave Desert AQMD", counties: ["San Bernardino", "Riverside"], website: "https://www.mdaqmd.ca.gov" },
  { id: "monterey-bay-ard", name: "Monterey Bay Air Resources District", counties: ["Monterey", "San Benito", "Santa Cruz"], website: "https://www.mbard.org" },
  { id: "north-coast-unified-aqmd", name: "North Coast Unified AQMD", counties: ["Del Norte", "Humboldt", "Trinity"], website: "https://www.ncuaqmd.org" },
  { id: "northern-sierra-aqmd", name: "Northern Sierra AQMD", counties: ["Nevada", "Plumas", "Sierra"], website: "https://www.myairdistrict.com" },
  { id: "northern-sonoma-county-apcd", name: "Northern Sonoma County APCD", counties: ["Sonoma"], website: "https://www.nosocoair.ca.gov" },
  { id: "placer-county-apcd", name: "Placer County APCD", counties: ["Placer"], website: "https://www.placer.ca.gov" },
  { id: "sacramento-metro-aqmd", name: "Sacramento Metropolitan AQMD", counties: ["Sacramento"], website: "https://www.airquality.org" },
  { id: "san-diego-county-apcd", name: "San Diego County APCD", counties: ["San Diego"], website: "https://www.sdapcd.org" },
  { id: "san-joaquin-valley-apcd", name: "San Joaquin Valley Unified APCD", counties: ["Fresno", "Kings", "Madera", "Merced", "San Joaquin", "Stanislaus", "Tulare", "Kern"], website: "https://www.valleyair.org" },
  { id: "san-luis-obispo-county-apcd", name: "San Luis Obispo County APCD", counties: ["San Luis Obispo"], website: "https://www.slocleanair.org" },
  { id: "santa-barbara-county-apcd", name: "Santa Barbara County APCD", counties: ["Santa Barbara"], website: "https://www.ourair.org" },
  { id: "shasta-county-aqmd", name: "Shasta County AQMD", counties: ["Shasta"], website: "https://www.co.shasta.ca.us" },
  { id: "siskiyou-county-apcd", name: "Siskiyou County APCD", counties: ["Siskiyou"], website: "https://www.co.siskiyou.ca.us" },
  { id: "south-coast-aqmd", name: "South Coast AQMD", counties: ["Orange", "Los Angeles", "San Bernardino", "Riverside"], website: "https://www.aqmd.gov" },
  { id: "tehama-county-apcd", name: "Tehama County APCD", counties: ["Tehama"], website: "https://www.tehcoapcd.net" },
  { id: "tuolumne-county-apcd", name: "Tuolumne County APCD", counties: ["Tuolumne"], website: "https://www.co.tuolumne.ca.us" },
  { id: "ventura-county-apcd", name: "Ventura County APCD", counties: ["Ventura"], website: "https://www.vcapcd.org" },
  { id: "yolo-solano-aqmd", name: "Yolo-Solano AQMD", counties: ["Yolo", "Solano"], website: "https://www.ysaqmd.org" },
];

// Counties split across more than one air district (resolution needs geometry).
export const SPLIT_AIR_COUNTIES: ReadonlySet<string> = new Set(["Kern", "Los Angeles", "Riverside", "San Bernardino", "Solano", "Sonoma"]);

export const REGIONAL_WATER_BOARDS: RegionalWaterBoard[] = [
  { id: "region-1-north-coast", name: "North Coast Regional Water Quality Control Board", coverage: "Del Norte, Humboldt, Mendocino, and north-coastal watersheds" },
  { id: "region-2-san-francisco-bay", name: "San Francisco Bay Regional Water Quality Control Board", coverage: "San Francisco Bay Area watersheds" },
  { id: "region-3-central-coast", name: "Central Coast Regional Water Quality Control Board", coverage: "Santa Cruz to Ventura coastal watersheds" },
  { id: "region-4-los-angeles", name: "Los Angeles Regional Water Quality Control Board", coverage: "Los Angeles and Ventura coastal watersheds" },
  { id: "region-5-central-valley", name: "Central Valley Regional Water Quality Control Board", coverage: "Sacramento and San Joaquin valleys (largest region)" },
  { id: "region-6-lahontan", name: "Lahontan Regional Water Quality Control Board", coverage: "Eastern Sierra / Lake Tahoe / high desert" },
  { id: "region-7-colorado-river-basin", name: "Colorado River Basin Regional Water Quality Control Board", coverage: "Imperial and eastern desert watersheds" },
  { id: "region-8-santa-ana", name: "Santa Ana Regional Water Quality Control Board", coverage: "Santa Ana River watershed (Orange + parts of Riverside/San Bernardino)" },
  { id: "region-9-san-diego", name: "San Diego Regional Water Quality Control Board", coverage: "San Diego region watersheds" },
];

export type AirDistrictResolution = {
  county: string;
  districts: AirDistrict[];
  // True when the county spans multiple air districts and a precise answer needs
  // sub-county geometry (lat/long), not just the county name. The caller must
  // not pick one arbitrarily — that would be a guess.
  needsGeometry: boolean;
};

// Resolve the candidate air district(s) for a county. Returns every district that
// covers the county; needsGeometry is true when there is more than one (the five
// known sub-county splits). Unknown county -> empty (never a guess).
export function resolveAirDistrict(county: string): AirDistrictResolution {
  const norm = county.trim().toLowerCase();
  const districts = AIR_DISTRICTS.filter((d) => d.counties.some((c) => c.toLowerCase() === norm));
  return { county, districts, needsGeometry: districts.length > 1 };
}

// County -> Regional Water Quality Control Board region number(s).
// Source: State Water Resources Control Board "The Nine Regional Water Quality
// Control Boards in California" fact sheet (waterboards.ca.gov). Region lines are
// HYDROLOGIC (watershed), not county lines, so many counties span multiple
// regions; those are listed with every region the fact sheet names. The legally
// controlling boundary is the GIS layer — multi-region counties therefore need
// sub-county geometry to pin a single board (needsGeometry below).
export const COUNTY_WATER_REGIONS: Readonly<Record<string, number[]>> = {
  Alameda: [2], Alpine: [6], Amador: [5], Butte: [5], Calaveras: [5], Colusa: [5],
  "Contra Costa": [2, 5], "Del Norte": [1], "El Dorado": [5, 6], Fresno: [5], Glenn: [1, 5],
  Humboldt: [1], Imperial: [7, 9], Inyo: [6], Kern: [3, 5, 6], Kings: [5], Lake: [1, 5],
  Lassen: [5, 6], "Los Angeles": [4, 6], Madera: [5], Marin: [1, 2], Mariposa: [5],
  Mendocino: [1], Merced: [5], Modoc: [1, 5, 6], Mono: [6], Monterey: [3], Napa: [2, 5],
  Nevada: [5, 6], Orange: [8], Placer: [5, 6], Plumas: [5], Riverside: [7, 8, 9],
  Sacramento: [5], "San Benito": [3, 5], "San Bernardino": [6, 7, 8], "San Diego": [7, 9],
  "San Francisco": [2], "San Joaquin": [5], "San Luis Obispo": [3, 5], "San Mateo": [2, 3],
  "Santa Barbara": [3, 4], "Santa Clara": [2, 3], "Santa Cruz": [3], Shasta: [5], Sierra: [5, 6],
  Siskiyou: [1, 5], Solano: [2, 5], Sonoma: [1, 2], Stanislaus: [5], Sutter: [5], Tehama: [5],
  Trinity: [1], Tulare: [5], Tuolumne: [5], Ventura: [3, 4], Yolo: [5], Yuba: [5],
};

const BOARD_BY_REGION = new Map<number, RegionalWaterBoard>(
  REGIONAL_WATER_BOARDS.map((b) => [Number(b.id.split("-")[1]), b]),
);

export type WaterBoardResolution = {
  county: string;
  boards: RegionalWaterBoard[];
  // True when the county spans more than one region — a single board can't be
  // chosen on the county name alone (needs sub-county watershed geometry).
  needsGeometry: boolean;
};

// Resolve the regional water board(s) for a county. Returns every board whose
// region covers the county; needsGeometry is true when more than one does.
// Unknown county -> empty (never a guess).
export function resolveWaterBoard(county: string): WaterBoardResolution {
  const norm = county.trim().replace(/\s+county$/i, "");
  const key = Object.keys(COUNTY_WATER_REGIONS).find((c) => c.toLowerCase() === norm.toLowerCase());
  const regions = key ? COUNTY_WATER_REGIONS[key] : [];
  const boards = regions.map((n) => BOARD_BY_REGION.get(n)).filter((b): b is RegionalWaterBoard => !!b);
  return { county, boards, needsGeometry: boards.length > 1 };
}

File: src/research_core/jurisdiction_registry.py
from __future__ import annotations

from dataclasses import dataclass

from research_core.jurisdiction_skills import normalize_county_name


@dataclass(frozen=True)
class AirDistrict:
    id: str
    name: str
    counties: tuple[str, ...]
    website: str


@dataclass(frozen=True)
class RegionalWaterBoard:
    id: str
    name: str
    coverage: str


@dataclass(frozen=True)
class AirDistrictResolution:
    county: str
    districts: tuple[AirDistrict, ...]
    needs_geometry: bool


@dataclass(frozen=True)
class WaterBoardResolution:
    county: str
    boards: tuple[RegionalWaterBoard, ...]
    needs_geometry: bool


AIR_DISTRICTS: tuple[AirDistrict, ...] = (
    AirDistrict("amador-county-apcd", "Amador County APCD", ("Amador",), "https://www.amadorgov.org"),
    AirDistrict("antelope-valley-aqmd", "Antelope Valley AQMD", ("Los Angeles",), "https://www.avaqmd.ca.gov"),
    AirDistrict("bay-area-aqmd", "Bay Area AQMD", ("Alameda", "Contra Costa", "Marin", "Napa", "San Francisco", "San Mateo", "Santa Clara", "Solano", "Sonoma"), "https://www.baaqmd.gov"),
    AirDistrict("butte-county-aqmd", "Butte County AQMD", ("Butte",), "https://www.bcaqmd.org"),
    AirDistrict("calaveras-county-apcd", "Calaveras County APCD", ("Calaveras",), "https://ema.calaverasgov.us"),
    AirDistrict("colusa-county-apcd", "Colusa County APCD", ("Colusa",), "https://www.countyofcolusaca.gov"),
    AirDistrict("eastern-kern-apcd", "Eastern Kern APCD", ("Kern",), "https://www.kernair.org"),
    AirDistrict("el-dorado-county-aqmd", "El Dorado County AQMD", ("El Dorado",), "https://www.eldoradocounty.ca.gov"),
    AirDistrict("feather-river-aqmd", "Feather River AQMD", ("Sutter", "Yuba"), "https://www.fraqmd.org"),
    AirDistrict("glenn-county-apcd", "Glenn County APCD", ("Glenn",), "https://www.countyofglenn.net"),
    AirDistrict("great-basin-unified-apcd", "Great Basin Unified APCD", ("Alpine", "Inyo", "Mono"), "https://www.gbuapcd.org"),
    AirDistrict("imperial-county-apcd", "Imperial County APCD", ("Imperial",), "https://www.imperialcounty.net"),
    AirDistrict("lake-county-aqmd", "Lake County AQMD", ("Lake",), "https://www.lcaqmd.net"),
    AirDistrict("lassen-county-apcd", "Lassen County APCD", ("Lassen",), "https://www.lassenair.org"),
    AirDistrict("mariposa-county-apcd", "Mariposa County APCD", ("Mariposa",), "https://www.mariposacounty.org"),
    AirDistrict("mendocino-county-aqmd", "Mendocino County AQMD", ("Mendocino",), "https://www.co.mendocino.ca.us"),
    AirDistrict("modoc-county-apcd", "Modoc County APCD", ("Modoc",), "https://www.co.modoc.ca.us"),
    AirDistrict("mojave-desert-aqmd", "Mojave Desert AQMD", ("San Bernardino", "Riverside"), "https://www.mdaqmd.ca.gov"),
    AirDistrict("monterey-bay-ard", "Monterey Bay Air Resources District", ("Monterey", "San Benito", "Santa Cruz"), "https://www.mbard.org"),
    AirDistrict("north-coast-unified-aqmd", "North Coast Unified AQMD", ("Del Norte", "Humboldt", "Trinity"), "https://www.ncuaqmd.org"),
    AirDistrict("northern-sierra-aqmd", "Northern Sierra AQMD", ("Nevada", "Plumas", "Sierra"), "https://www.myairdistrict.com"),
    AirDistrict("northern-sonoma-county-apcd", "Northern Sonoma County APCD", ("Sonoma",), "https://www.nosocoair.ca.gov"),
    AirDistrict("placer-county-apcd", "Placer County APCD", ("Placer",), "https://www.placer.ca.gov"),
    AirDistrict("sacramento-metro-aqmd", "Sacramento Metropolitan AQMD", ("Sacramento",), "https://www.airquality.org"),
    AirDistrict("san-diego-county-apcd", "San Diego County APCD", ("San Diego",), "https://www.sdapcd.org"),
    AirDistrict("san-joaquin-valley-apcd", "San Joaquin Valley Unified APCD", ("Fresno", "Kings", "Madera", "Merced", "San Joaquin", "Stanislaus", "Tulare", "Kern"), "https://www.valleyair.org"),
    AirDistrict("san-luis-obispo-county-apcd", "San Luis Obispo County APCD", ("San Luis Obispo",), "https://www.slocleanair.org"),
    AirDistrict("santa-barbara-county-apcd", "Santa Barbara County APCD", ("Santa Barbara",), "https://www.ourair.org"),
    AirDistrict("shasta-county-aqmd", "Shasta County AQMD", ("Shasta",), "https://www.co.shasta.ca.us"),
    AirDistrict("siskiyou-county-apcd", "Siskiyou County APCD", ("Siskiyou",), "https://www.co.siskiyou.ca.us"),
    AirDistrict("south-coast-aqmd", "South Coast AQMD", ("Orange", "Los Angeles", "San Bernardino", "Riverside"), "https://www.aqmd.gov"),
    AirDistrict("tehama-county-apcd", "Tehama County APCD", ("Tehama",), "https://www.tehcoapcd.net"),
    AirDistrict("tuolumne-county-apcd", "Tuolumne County APCD", ("Tuolumne",), "https://www.co.tuolumne.ca.us"),
    AirDistrict("ventura-county-apcd", "Ventura County APCD", ("Ventura",), "https://www.vcapcd.org"),
    AirDistrict("yolo-solano-aqmd", "Yolo-Solano AQMD", ("Yolo", "Solano"), "https://www.ysaqmd.org"),
)


SPLIT_AIR_COUNTIES: frozenset[str] = frozenset(
    ["Kern", "Los Angeles", "Riverside", "San Bernardino", "Solano", "Sonoma"]
)


REGIONAL_WATER_BOARDS: tuple[RegionalWaterBoard, ...] = (
    RegionalWaterBoard("region-1-north-coast", "North Coast Regional Water Quality Control Board", "Del Norte, Humboldt, Mendocino, and north-coastal watersheds"),
    RegionalWaterBoard("region-2-san-francisco-bay", "San Francisco Bay Regional Water Quality Control Board", "San Francisco Bay Area watersheds"),
    RegionalWaterBoard("region-3-central-coast", "Central Coast Regional Water Quality Control Board", "Santa Cruz to Ventura coastal watersheds"),
    RegionalWaterBoard("region-4-los-angeles", "Los Angeles Regional Water Quality Control Board", "Los Angeles and Ventura coastal watersheds"),
    RegionalWaterBoard("region-5-central-valley", "Central Valley Regional Water Quality Control Board", "Sacramento and San Joaquin valleys (largest region)"),
    RegionalWaterBoard("region-6-lahontan", "Lahontan Regional Water Quality Control Board", "Eastern Sierra / Lake Tahoe / high desert"),
    RegionalWaterBoard("region-7-colorado-river-basin", "Colorado River Basin Regional Water Quality Control Board", "Imperial and eastern desert watersheds"),
    RegionalWaterBoard("region-8-santa-ana", "Santa Ana Regional Water Quality Control Board", "Santa Ana River watershed (Orange + parts of Riverside/San Bernardino)"),
    RegionalWaterBoard("region-9-san-diego", "San Diego Regional Water Quality Control Board", "San Diego region watersheds"),
)


COUNTY_WATER_REGIONS: dict[str, tuple[int, ...]] = {
    "Alameda": (2,),
    "Alpine": (6,),
    "Amador": (5,),
    "Butte": (5,),
    "Calaveras": (5,),
    "Colusa": (5,),
    "Contra Costa": (2, 5),
    "Del Norte": (1,),
    "El Dorado": (5, 6),
    "Fresno": (5,),
    "Glenn": (1, 5),
    "Humboldt": (1,),
    "Imperial": (7, 9),
    "Inyo": (6,),
    "Kern": (3, 5, 6),
    "Kings": (5,),
    "Lake": (1, 5),
    "Lassen": (5, 6),
    "Los Angeles": (4, 6),
    "Madera": (5,),
    "Marin": (1, 2),
    "Mariposa": (5,),
    "Mendocino": (1,),
    "Merced": (5,),
    "Modoc": (1, 5, 6),
    "Mono": (6,),
    "Monterey": (3,),
    "Napa": (2, 5),
    "Nevada": (5, 6),
    "Orange": (8,),
    "Placer": (5, 6),
    "Plumas": (5,),
    "Riverside": (7, 8, 9),
    "Sacramento": (5,),
    "San Benito": (3, 5),
    "San Bernardino": (6, 7, 8),
    "San Diego": (7, 9),
    "San Francisco": (2,),
    "San Joaquin": (5,),
    "San Luis Obispo": (3, 5),
    "San Mateo": (2, 3),
    "Santa Barbara": (3, 4),
    "Santa Clara": (2, 3),
    "Santa Cruz": (3,),
    "Shasta": (5,),
    "Sierra": (5, 6),
    "Siskiyou": (1, 5),
    "Solano": (2, 5),
    "Sonoma": (1, 2),
    "Stanislaus": (5,),
    "Sutter": (5,),
    "Tehama": (5,),
    "Trinity": (1,),
    "Tulare": (5,),
    "Tuolumne": (5,),
    "Ventura": (3, 4),
    "Yolo": (5,),
    "Yuba": (5,),
}


BOARD_BY_REGION: dict[int, RegionalWaterBoard] = {
    int(board.id.split("-")[1]): board for board in REGIONAL_WATER_BOARDS
}


def resolve_air_district(county: str) -> AirDistrictResolution:
    norm = normalize_county_name(county).lower()
    districts = tuple(
        district
        for district in AIR_DISTRICTS
        if any(c.lower() == norm for c in district.counties)
    )
    return AirDistrictResolution(
        county=county,
        districts=districts,
        needs_geometry=len(districts) > 1,
    )


def resolve_water_board(county: str) -> WaterBoardResolution:
    norm = normalize_county_name(county)
    key = next(
        (c for c in COUNTY_WATER_REGIONS if c.lower() == norm.lower()),
        None,
    )
    regions = COUNTY_WATER_REGIONS[key] if key else ()
    boards = tuple(
        BOARD_BY_REGION[region]
        for region in regions
        if region in BOARD_BY_REGION
    )
    return WaterBoardResolution(
        county=county,
        boards=boards,
        needs_geometry=len(boards) > 1,
    )
